### PR TITLE
[TTS]Diffsinger opencpop baseline

### LINKED ISSUE #2821: [TTS] DiffSinger
- https://github.com/PaddlePaddle/PaddleSpeech/pull/3005
- https://github.com/PaddlePaddle/PaddleSpeech/pull/2902
- https://github.com/PaddlePaddle/PaddleSpeech/pull/2868
- https://github.com/PaddlePaddle/PaddleSpeech/pull/2834
- https://github.com/PaddlePaddle/PaddleSpeech/pull/2832

### PR DESCRIPTION
**Pull request by lym0302**
PR types
new feature

PR changes
add Diffsinger opencpop baseline （fft training）
 
Describe
add Diffsinger opencpop baseline （fft training）

fix https://github.com/PaddlePaddle/PaddleSpeech/issues/2821

### CODE REVIEW

File: examples/opencpop/svs1/run.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+set -e
+source path.sh
+
+gpus=4,5
```
Comment (yt605155624) [inline examples/opencpop/svs1/run.sh]: 默认改成 0,1

File: paddlespeech/t2s/datasets/get_feats.py [hunk 2]
```diff
@@ -166,6 +167,8 @@ def get_pitch(self,
         f0 = self._calculate_f0(wav, use_continuous_f0, use_log_f0)
         if use_token_averaged_f0 and duration is not None:
             f0 = self._average_by_duration(f0, duration)
+        else:
```
Comment (yt605155624) [inline paddlespeech/t2s/datasets/get_feats.py:170]: 这样改的话，除了时间维度的长度，其他是否和 if 分支保持一致？如 dim 的个数，dim 的顺序？（[T, channel] or [channel, T]）

File: paddlespeech/t2s/datasets/get_feats.py [hunk 3]
```diff
@@ -224,6 +227,8 @@ def get_energy(self,
         energy = self._calculate_energy(wav)
         if use_token_averaged_energy and duration is not None:
             energy = self._average_by_duration(energy, duration)
+        else:
+            energy = np.expand_dims(np.array(energy), 0).T
```
Comment (yt605155624) [inline paddlespeech/t2s/datasets/get_feats.py:231]: 问题同 pitch

File: paddlespeech/t2s/datasets/preprocess_utils.py [hunk 4]
```diff
@@ -40,6 +42,58 @@ def get_phn_dur(file_name):
     f.close()
     return sentence, speaker_set
 
+def note2midi(notes):
```
Comment (yt605155624) [inline paddlespeech/t2s/datasets/preprocess_utils.py]: 建议加上 typehint 便于理解

File: paddlespeech/t2s/datasets/preprocess_utils.py [hunk 5]
```diff
@@ -40,6 +42,58 @@ def get_phn_dur(file_name):
...
+    for note in notes:
+        if note == 'rest':
+            midi = 0
+        else:
+            midi = librosa.note_to_midi(note.split("/")[0])
+        midis.append(midi)
+
+    return midis
+
+def time2frame(times, sample_rate: int=24000, n_shift: int=128,):
```
Comment (yt605155624) [inline paddlespeech/t2s/datasets/preprocess_utils.py]: times 的 typehint?

File: paddlespeech/t2s/exps/diffsinger/preprocess.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+from paddlespeech.t2s.datasets.get_feats import LogMelFBank
+from paddlespeech.t2s.datasets.get_feats import Pitch
+from paddlespeech.t2s.datasets.preprocess_utils import compare_duration_and_mel_length
+from paddlespeech.t2s.datasets.preprocess_utils import get_input_token
+from paddlespeech.t2s.datasets.preprocess_utils import get_sentences_svs
+from paddlespeech.t2s.datasets.preprocess_utils import get_spk_id_map
+from paddlespeech.t2s.datasets.preprocess_utils import merge_silence
+from paddlespeech.t2s.utils import str2bool
+
+ALL_SHENGMU = ['zh', 'ch', 'sh', 'b', 'p', 'm', 'f', 'd', 't', 'n', 'l', 'g', 'k', 'h', 'j',
```
Comment (yt605155624) [inline paddlespeech/t2s/exps/diffsinger/preprocess.py]: 用英文 INITIALS 和 FINALS

File: paddlespeech/t2s/exps/syn_utils.py [hunk 7]
```diff
@@ -23,6 +23,7 @@
 import numpy as np
 import onnxruntime as ort
 import paddle
+import yaml
```
Comment (yt605155624) [inline paddlespeech/t2s/exps/syn_utils.py]: 多余？

File: paddlespeech/t2s/models/diffsinger/diffsinger.py [hunk 8]
```diff
@@ -0,0 +1,1286 @@
...
+                is_inference=True,
+                alpha=alpha,
+                spk_emb=spk_emb,
+                spk_id=spk_id,
+                tone_id=tone_id)
+
+        return outs[0], d_outs[0], p_outs[0], e_outs[0]
+
+
+class FastSpeech2MIDILoss(nn.Layer):
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/diffsinger.py]: 是否可以直接复用 FastSpeech2Loss

File: paddlespeech/t2s/models/diffsinger/diffsinger.py [hunk 9]
```diff
@@ -0,0 +1,1286 @@
...
+
+    Args:
+    
+    Returns:
+
+    """
+
+    def __init__(
+            self,
+            fs2_config,
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/diffsinger.py]: 这里我还是倾向于和其他模型保持风格一致，原因如下:
1. 不知道有哪些参数，也不知道参数的类型，
2. 有的参数可能可以有多种类型，而 yaml 里面只能写一种，比如 diffusion 的 beta_schedule yaml 里面选择了 "squaredcos_cap_v2"，是否有其他可选项？
3. 需要在 `__init__` 函数下写上详细的注释，说明每种参数的含义、Tensor Shape 和可选项等
4. 如果觉得写上的话参数太多，可以把 FastSpeech2MIDI 这个类放到其他位置，diffusion.py 只写 diffusion 的组网

File: paddlespeech/t2s/modules/transformer/encoder.py [hunk 10]
```diff
@@ -390,12 +390,18 @@ def __init__(
             padding_idx=padding_idx,
             encoder_type="transformer")
 
-    def forward(self, xs, masks):
+    def forward(self, xs, masks, note_emb=None, note_dur_emb=None, is_slur_emb=None, scale=16):
```
Comment (yt605155624) [inline paddlespeech/t2s/modules/transformer/encoder.py]: 参数多了建议加上 typehint

File: examples/opencpop/svs1/local/train.sh [hunk 11]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Comment (yt605155624) [inline examples/opencpop/svs1/local/train.sh:1]: 这个其实可以直接软连接 csmsc/tts3/local/train.sh (除了 ngpu 不一样)，可能等后期再改

File: paddlespeech/t2s/models/diffsinger/diffsinger.py [hunk 12]
```diff
@@ -0,0 +1,1286 @@
...
+from paddlespeech.t2s.modules.transformer.encoder import CNNPostnet
+from paddlespeech.t2s.modules.transformer.encoder import ConformerEncoder
+from paddlespeech.t2s.modules.transformer.encoder import TransformerEncoder
+
+
+class FastSpeech2MIDI(FastSpeech2):
+    """The Fastspeech2 module of DiffSinger.
+    """
+
+    def __init__(
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/diffsinger.py:42]: 最大程度复用父类的 `__init__` 函数，fs2 相关参数可以以字典的形式传入，此处 fs2_config 直接给父类的 `__init__`

File: paddlespeech/t2s/models/diffsinger/fastspeech2midi.py [hunk 13]
```diff
@@ -0,0 +1,625 @@
...
+                ilens,
+                is_inference=True,
+                alpha=alpha,
+                spk_emb=spk_emb,
+                spk_id=spk_id, )
+
+        return outs[0], d_outs[0], p_outs[0], e_outs[0]
+
+
+class FastSpeech2MIDILoss(nn.Layer):
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/fastspeech2midi.py:476]: 可以继承自 FastSpeech2Loss, 直接使用父类的 `__init__`

File: paddlespeech/t2s/models/diffsinger/diffsinger.py [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+            fastspeech2_params (Dict[str, Any]): Parameter dict for fastspeech2 module.
+            denoiser_params (Dict[str, Any]): Parameter dict for dinoiser module.
+            diffusion_params (Dict[str, Any]): Parameter dict for diffusion module.
+        """
+        assert check_argument_types()
+        super().__init__()
+        self.fs2 = FastSpeech2MIDI(
+            idim=idim,
+            odim=odim,
+            fastspeech2_config=fastspeech2_params,
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/diffsinger.py:152]: 一个是 fastspeech2_config 一个是 fastspeech2_params 此处是否保持一致

File: paddlespeech/t2s/models/diffsinger/diffsinger_updater.py [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+
+
+class DiffSingerUpdater(StandardUpdater):
+    def __init__(
+            self,
+            model: Layer,
+            optimizers: Dict[str, Optimizer],
+            criterions: Dict[str, Layer],
+            dataloader: DataLoader,
+            fs2_train_start_steps: int=0,
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/diffsinger_updater.py:44]: 这个参数感觉不需要，fs2_train_start_steps 存在不是 0 的情况吗？

File: paddlespeech/t2s/models/diffsinger/fastspeech2midi.py [hunk 16]
```diff
@@ -0,0 +1,625 @@
...
+        ds = paddle.cast(durations, 'int64')
+        ps = pitch
+        es = energy
+        ys = speech
+        olens = speech_lengths
+        if spk_id is not None:
+            spk_id = paddle.cast(spk_id, 'int64')
+        # forward propagation
+        before_outs, after_outs, d_outs, p_outs, e_outs, spk_logits = self._forward(
+            xs,
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/fastspeech2midi.py:136]: 建议调用时加上形参，参数太多了，容易乱

File: paddlespeech/t2s/models/diffsinger/fastspeech2midi.py [hunk 17]
```diff
@@ -0,0 +1,625 @@
...
+            spk_emb = spk_emb.unsqueeze(0)
+
+        if use_teacher_forcing:
+            # use groundtruth of duration, pitch, and energy
+            ds = d.unsqueeze(0) if d is not None else None
+            ps = p.unsqueeze(0) if p is not None else None
+            es = e.unsqueeze(0) if e is not None else None
+
+            # (1, L, odim)
+            _, outs, d_outs, p_outs, e_outs, _ = self._forward(
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/fastspeech2midi.py:448]: 调用时建议加上形参

File: paddlespeech/t2s/models/diffsinger/fastspeech2midi.py [hunk 18]
```diff
@@ -0,0 +1,625 @@
...
+                ilens,
+                ds=ds,
+                ps=ps,
+                es=es,
+                spk_emb=spk_emb,
+                spk_id=spk_id,
+                is_inference=True)
+        else:
+            # (1, L, odim)
+            _, outs, d_outs, p_outs, e_outs, _ = self._forward(
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/fastspeech2midi.py:462]: 调用时建议加上形参

File: paddlespeech/t2s/models/diffsinger/fastspeech2midi.py [hunk 19]
```diff
@@ -0,0 +1,625 @@
...
+        is_slur = paddle.cast(is_slur, 'int64')
+        ilens = paddle.cast(text_lengths, 'int64')
+        olens = paddle.cast(speech_lengths, 'int64')
+
+        if spk_emb is not None:
+            spk_emb = spk_emb.unsqueeze(0)
+
+        # (1, L, odim)
+        # use *_ to avoid bug in dygraph to static graph    
+        hs, h_masks = self._forward(
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/fastspeech2midi.py:369]: 调用时建议加上形参

File: paddlespeech/t2s/models/diffsinger/fastspeech2midi.py [hunk 20]
```diff
@@ -0,0 +1,625 @@
...
+        is_slur = paddle.cast(is_slur, 'int64').unsqueeze(0)
+        # setup batch axis
+        ilens = paddle.shape(xs)[1]
+
+        if spk_emb is not None:
+            spk_emb = spk_emb.unsqueeze(0)
+
+        # (1, L, odim)
+        # use *_ to avoid bug in dygraph to static graph    
+        hs, _ = self._forward(
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/fastspeech2midi.py:328]: 调用时建议加上形参

File: paddlespeech/t2s/models/diffsinger/diffsinger_updater.py [hunk 21]
```diff
@@ -0,0 +1,242 @@
...
+
+            loss_ds = l1_loss_ds
+
+            self.optimizer_ds.clear_grad()
+            loss_ds.backward()
+            self.optimizer_ds.step()
+
+            report("train/loss_ds", float(loss_ds))
+            report("train/l1_loss_ds", float(l1_loss_ds))
+            losses_dict["l1_loss_ds"] = float(l1_loss_ds)
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/diffsinger_updater.py:172]: 这两个 loss 一样的话是否需要 report 两遍？

File: paddlespeech/t2s/models/diffsinger/diffsinger.py [hunk 22]
```diff
@@ -0,0 +1,356 @@
...
+        return mel[0]
+
+
+class DiffSingerInference(nn.Layer):
+    def __init__(self, normalizer, model):
+        super().__init__()
+        self.normalizer = normalizer
+        self.acoustic_model = model
+
+    def forward(self, text, note, note_dur, is_slur, get_mel_fs2: bool=False):
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/diffsinger.py:282]: typehint

File: paddlespeech/t2s/exps/diffsinger/train.py [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+from paddlespeech.t2s.models.diffsinger import DiffusionLoss
+from paddlespeech.t2s.models.diffsinger.fastspeech2midi import FastSpeech2MIDILoss
+from paddlespeech.t2s.training.extensions.snapshot import Snapshot
+from paddlespeech.t2s.training.extensions.visualizer import VisualDL
+from paddlespeech.t2s.training.optimizer import build_optimizers
+from paddlespeech.t2s.training.seeding import seed_everything
+from paddlespeech.t2s.training.trainer import Trainer
+from paddlespeech.t2s.utils import str2bool
+
+# from paddlespeech.t2s.models.fastspeech2 import FastSpeech2Loss
```
Comment (yt605155624) [inline paddlespeech/t2s/exps/diffsinger/train.py:48]: 可以删掉

File: paddlespeech/t2s/datasets/preprocess_utils.py [hunk 24]
```diff
@@ -41,6 +45,90 @@ def get_phn_dur(file_name):
...
+        sample_rate: int=24000,
+        n_shift: int=128, ):
+    '''
+    read label file
+    Args:
+        file_name (str or Path): path of gen_duration_from_textgrid.py's result
+        dataset (str): dataset name
+    Returns: 
+        Dict: the information of sentence, include [phone id (int)], [the frame of phone (int)], [note id (int)], [note duration (float)], [is slur (int)], text(str), speaker name (str)
+        tunple: speaker name
```
Comment (yt605155624) [inline paddlespeech/t2s/datasets/preprocess_utils.py:104]: tunple 拼写错误

File: paddlespeech/t2s/exps/diffsinger/train.py [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+
+    with open(args.config) as f:
+        config = CfgNode(yaml.safe_load(f))
+
+    print("========Args========")
+    print(yaml.safe_dump(vars(args)))
+    print("========Config========")
+    print(config)
+    print(
+        f"master see the word size: {dist.get_world_size()}, from pid: {os.getpid()}"
```
Comment (yt605155624) [inline paddlespeech/t2s/exps/diffsinger/train.py:234]: world

File: paddlespeech/t2s/models/diffsinger/diffsinger.py [hunk 26]
```diff
@@ -0,0 +1,356 @@
...
+        Returns:
+            _type_: _description_
+        """
+        mel_fs2, _, _, _ = self.fs2.inference(text, note, note_dur, is_slur)
+        if get_mel_fs2:
+            return mel_fs2
+        mel_fs2 = mel_fs2.unsqueeze(0).transpose((0, 2, 1))
+        cond_fs2 = self.fs2.encoder_infer(text, note, note_dur, is_slur)
+        cond_fs2 = cond_fs2.transpose((0, 2, 1))
+        mel, _ = self.diffusion(mel_fs2, cond_fs2)
```
Comment (yt605155624) [inline paddlespeech/t2s/models/diffsinger/diffsinger.py:271]: 此处是否应该调用 self.diffusion.inference() , 如果是的话 应该加一个 num_inference_steps 参数控制下步数，默认的 1000 太大了
